File: WinPaletter/GUI/Elements/AdvancedControls/TextTranslationItem.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D

Namespace UI.Controllers

    <ToolboxItem(False)> Public Class TextTranslationItem : Inherits ContainerControl

        Sub New()
            DoubleBuffered = True
            SetStyle(ControlStyles.SupportsTransparentBackColor, True)
            BackColor = Color.Transparent
        End Sub

        Protected Overrides ReadOnly Property CreateParams As CreateParams
            Get
                Dim cp As CreateParams = MyBase.CreateParams
                cp.ExStyle = cp.ExStyle Or &H20
                Return cp
            End Get
        End Property

#Region "Properties"
        Public Property Image As Image

        Private _TextAlign As ContentAlignment = ContentAlignment.MiddleCenter
        Public Property TextAlign As ContentAlignment
            Get
                Return _TextAlign
            End Get
            Set(value As ContentAlignment)
                _TextAlign = value
                Invalidate()
            End Set
        End Property

        Private _ImageAlign As ContentAlignment = ContentAlignment.MiddleCenter
        Public Property ImageAlign As ContentAlignment
            Get
                Return _ImageAlign
            End Get
            Set(value As ContentAlignment)
                _ImageAlign = value
                Invalidate()
            End Set
        End Property

        Private _Text As String
        <Browsable(True)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)>
        <EditorBrowsable(EditorBrowsableState.Always)>
        <Editor(GetType(System.ComponentModel.Design.MultilineStringEditor), GetType(System.Drawing.Design.UITypeEditor))>
        <Bindable(True)>
        Public Overrides Property Text As String
            Get
                Return _Text
            End Get
            Set(value As String)
                _Text = value
                Invalidate()
            End Set
        End Property

        Public Pressed As Boolean
        Public Property Text_English As String
        Public Property Tag_English As String

        Private _SearchHilight As String
        Public Property SearchHilight As String
            Get
                Return _SearchHilight
            End Get
            Set(value As String)
                _SearchHilight = value
                Refresh()
            End Set
        End Property

#End Region

#Region "Events"
        Protected Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
            MyBase.OnMouseDown(e)
            Pressed = True
            BringToFront()
            Focus()
            Invalidate()
        End Sub

        Private Sub Button_LostFocus(sender As Object, e As EventArgs) Handles Me.LostFocus
            Pressed = False : Invalidate()
        End Sub

        Private Sub ItemSelection_GotFocus(sender As Object, e As EventArgs) Handles Me.GotFocus
            Pressed = True
            BringToFront()
            Focus()
            Invalidate()
        End Sub


#End Region
        Protected Overrides Sub OnPaint(e As PaintEventArgs)
            Dim G As Graphics = e.Graphics
            G.SmoothingMode = SmoothingMode.HighQuality
            G.TextRenderingHint = My.RenderingHint
            DoubleBuffered = True
            Dim rect As New Rectangle(0, 0, Width - 1, Height - 1)
            Dim NotTranslatedColor As Color = My.Style.Colors.NotTranslatedColor

            If _SearchHilight IsNot Nothing AndAlso Not String.IsNullOrWhiteSpace(_SearchHilight) AndAlso Text.ToLower.Trim.Contains(_SearchHilight.ToLower.Trim) Then
                Using br As New SolidBrush(My.Style.Colors.SearchColor) : G.FillRectangle(br, rect) : End Using

            ElseIf Not String.IsNullOrWhiteSpace(Text) AndAlso Text.Trim = Text_English.Trim Then
                Using br As New SolidBrush(NotTranslatedColor) : G.FillRectangle(br, rect) : End Using

            ElseIf Tag IsNot Nothing AndAlso Not String.IsNullOrWhiteSpace(Tag.ToString) AndAlso Tag.ToString.Trim = Tag_English.Trim Then
                Using br As New SolidBrush(NotTranslatedColor) : G.FillRectangle(br, rect) : End Using

            End If

            If Pressed Then
                G.FillRectangle(New SolidBrush(My.Style.Colors.Back_Checked), rect)
                G.DrawRectangle(New Pen(If(My.Style.DarkMode, Color.White, Color.Black), 2) With {.DashStyle = DashStyle.Dot}, rect)
            Else
                G.DrawRectangle(New Pen(Color.FromArgb(100, If(My.Style.DarkMode, Color.White, Color.Black)), 1), rect)
            End If

#Region "Text and Image Render"
            Using ButtonString As New StringFormat() With {.Alignment = StringAlignment.Center, .LineAlignment = StringAlignment.Center}
                Dim RTL As Boolean = (RightToLeft = 1)
                If RTL Then ButtonString.FormatFlags = StringFormatFlags.DirectionRightToLeft

                Dim img As Bitmap = Nothing
                If Image IsNot Nothing Then
                    If Enabled Then
                        img = CType(Image.Clone, Bitmap)
                    Else
                        img = CType(Image.Clone, Bitmap).Grayscale
                    End If
                End If

                Dim imgX, imgY As Integer

                Try
                    If img IsNot Nothing Then imgX = CInt((Width - img.Width) / 2)
                Catch : End Try

                Try : If img IsNot Nothing Then imgY = CInt((Height - img.Height) / 2)
                Catch : End Try

                If img Is Nothing Then
                    Try
                        Using br As New SolidBrush(ForeColor), sf As StringFormat = TextAlign.ToStringFormat(RTL)
                            Dim r As New Rectangle(1, 0, Width, Height)
                            G.DrawString(Text, Font, br, r, sf)
                        End Using
                    Catch
                    End Try
                Else

                    Select Case ImageAlign
                        Case ContentAlignment.MiddleCenter
                            ButtonString.Alignment = StringAlignment.Center : ButtonString.LineAlignment = StringAlignment.Near
                            Dim alx As Integer = CInt((Height - (img.Height + 4 + Text.Measure(MyBase.Font).Height)) / 2)

                            Try : If img IsNot Nothing Then
                                    If Text = Nothing Then
                                        G.DrawImage(img.Clone, New Rectangle(imgX, imgY, img.Width, img.Height))
                                    Else
                                        G.DrawImage(img.Clone, New Rectangle(imgX, alx, img.Width, img.Height))
                                    End If
                                End If

                                Using br As New SolidBrush(ForeColor)
                                    Dim r As New Rectangle(0, alx + 9 + img.Height, Width, Height)
                                    G.DrawString(Text, Font, br, r, ButtonString)
                                End Using
                            Catch : End Try

                        Case ContentAlignment.MiddleLeft
                            Dim Rec As New Rectangle(imgY, imgY, img.Width, img.Height)
                            Dim Bo As Integer = imgY + img.Width + imgY - 5
                            Dim RecText As New Rectangle(Bo, imgY, Text.Measure(Font).Width + 15 - imgY, img.Height)
                            Dim u As Rectangle = Rectangle.Union(Rec, RecText)
                            u.X = (Width - u.Width) / 2
                            Dim innerSpace As Integer = RecText.Left - Rec.Right

                            If Not RTL Then
                                Rec.X = u.Left
                                RecText.X = u.Left + Rec.Width + innerSpace
                            Else
                                Rec.X = u.Right - Rec.Width
                                RecText.X = u.Right - RecText.Width - Rec.Width - innerSpace
                            End If

                            G.DrawImage(img.Clone, Rec)
                            Using br As New SolidBrush(ForeColor)
                                G.DrawString(Text, Font, br, RecText, ButtonString)
                            End Using

                        Case ContentAlignment.MiddleRight
                            Dim Rec As New Rectangle(imgY, imgY, img.Width, img.Height)
                            Dim Bo As Integer = imgY + img.Width + imgY - 5
                            Dim RecText As New Rectangle(Bo, imgY, Width - Bo, img.Height)
                            Dim u As Rectangle = Rectangle.Union(Rec, RecText)
                            Dim innerSpace As Integer = RecText.Left - Rec.Right

                            If Not RTL Then
                                Rec.X = u.Left
                                RecText.X = u.Left + Rec.Width + innerSpace
                            Else
                                Rec.X = u.Right - Rec.Width - 2
                                RecText.X = u.Right - RecText.Width - Rec.Width - innerSpace
                            End If

                            G.DrawImage(img.Clone, Rec)
                            Using br As New SolidBrush(ForeColor)
                                G.DrawString(Text, Font, br, RecText, ButtonString)
                            End Using
                    End Select
                End If
            End Using
#End Region

        End Sub
    End Class

End Namespace

File: WinPaletter/GUI/Elements/AdvancedControls/TextTranslationItem.cs
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Drawing.Design;
using System.Drawing.Drawing2D;
using System.Drawing.Text;
using System.Windows.Forms;

namespace WinPaletter.UI.Controllers
{
    [ToolboxItem(false)]
    public class TextTranslationItem : Control
    {
        public TextTranslationItem()
        {
            DoubleBuffered = true;
            SetStyle(ControlStyles.SupportsTransparentBackColor, true);
            BackColor = Color.Transparent;
        }

        #region Properties

        private Image _image;
        public Image Image
        {
            get => _image;
            set
            {
                _image = value;
                Invalidate();
            }
        }


        private ContentAlignment _TextAlign = ContentAlignment.MiddleCenter;
        public ContentAlignment TextAlign
        {
            get => _TextAlign;
            set
            {
                if (value != _TextAlign)
                {
                    _TextAlign = value;
                    Invalidate();
                }
            }
        }


        private ContentAlignment _ImageAlign = ContentAlignment.MiddleCenter;
        public ContentAlignment ImageAlign
        {
            get => _ImageAlign;
            set
            {
                if (value != _ImageAlign)
                {
                    _ImageAlign = value;
                    Invalidate();
                }
            }
        }


        private TextImageRelation _TextImageRelation = TextImageRelation.ImageBeforeText;
        public TextImageRelation TextImageRelation
        {
            get => _TextImageRelation;
            set
            {
                if (value != _TextImageRelation)
                {
                    _TextImageRelation = value;
                    Invalidate();
                }
            }
        }


        private string _Text;
        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        [Editor(typeof(MultilineStringEditor), typeof(UITypeEditor))]
        [Bindable(true)]
        public override string Text
        {
            get => _Text;
            set
            {
                if (value != _Text)
                {
                    _Text = value;
                    Invalidate();
                }
            }
        }

        public bool Pressed;
        public string Text_English { get; set; }
        public string Tag_English { get; set; }

        private string _SearchHighlight;
        public string SearchHighlight
        {
            get => _SearchHighlight;
            set
            {
                if (value != _SearchHighlight)
                {
                    _SearchHighlight = value;
                    Invalidate();
                }
            }
        }

        public new Color ForeColor { get; set; } = Color.White;

        #endregion

        #region Events/Overrides

        protected override void OnMouseDown(MouseEventArgs e)
        {
            Pressed = true;
            BringToFront();
            Focus();
            Invalidate();

            base.OnMouseDown(e);
        }

        protected override void OnLostFocus(EventArgs e)
        {
            Pressed = false;
            Invalidate();

            base.OnLostFocus(e);
        }

        protected override void OnGotFocus(EventArgs e)
        {
            Pressed = true;
            BringToFront();
            Focus();
            Invalidate();

            base.OnGotFocus(e);
        }

        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);

            _image?.Dispose();
        }

        #endregion

        protected override void OnPaintBackground(PaintEventArgs pevent)
        {
            //Leave it empty to make control background transparent
            base.OnPaintBackground(pevent);
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics G = e.Graphics;
            G.SmoothingMode = SmoothingMode.HighQuality;
            G.TextRenderingHint = DesignMode ? TextRenderingHint.ClearTypeGridFit : Program.Style.TextRenderingHint;

            Rectangle rect = new(0, 0, Width - 1, Height - 1);
            Rectangle PaddingRect = Rectangle.FromLTRB(Padding.Left, Padding.Top, Padding.Right, Padding.Bottom);
            Rectangle TextAndImageRect = new(rect.X + PaddingRect.X, rect.Y + PaddingRect.Y, Width - 1 - PaddingRect.Width * 2 - rect.X * 2, Height - 1 - PaddingRect.Height * 2 - rect.Y * 2);

            if (_SearchHighlight is not null && !string.IsNullOrWhiteSpace(_SearchHighlight) && Text.ToLower().Trim().Contains(_SearchHighlight.ToLower().Trim()))
            {
                G.FillRoundedRect(Program.Style.Schemes.Tertiary.Brushes.Back_Checked_Hover, rect);
                G.DrawRoundedRectBeveled(Program.Style.Schemes.Tertiary.Pens.Line_Checked_Hover, rect);
            }

            else if (!string.IsNullOrWhiteSpace(Text) && (Text.Trim() ?? string.Empty) == (Text_English.Trim() ?? string.Empty))
            {
                G.FillRoundedRect(Program.Style.Schemes.Secondary.Brushes.Back_Checked, rect);
                G.DrawRoundedRectBeveled(Program.Style.Schemes.Secondary.Pens.Line_Checked, rect);
            }

            else if (Tag is not null && !string.IsNullOrWhiteSpace(Tag.ToString()) && (Tag.ToString().Trim() ?? string.Empty) == (Tag_English.Trim() ?? string.Empty))
            {
                G.FillRoundedRect(Program.Style.Schemes.Secondary.Brushes.Back_Checked, rect);
                G.DrawRoundedRectBeveled(Program.Style.Schemes.Secondary.Pens.Line_Checked, rect);
            }

            else
            {
                using (SolidBrush br = new(Program.Style.Schemes.Secondary.Colors.Back(0)))
                using (Pen P = new(Program.Style.Schemes.Secondary.Colors.Line(0)))
                {
                    G.FillRoundedRect(br, rect);
                    G.DrawRoundedRectBeveled(P, rect);
                }
            }

            if (Pressed)
            {
                G.FillRoundedRect(Program.Style.Schemes.Main.Brushes.Back_Checked, rect);
                G.DrawRoundedRectBeveled(Program.Style.Schemes.Main.Pens.Line_Checked, rect);
            }

            #region Text and Image Render

            using (StringFormat sf = TextAlign.ToStringFormat((int)base.RightToLeft == 1))
            using (SolidBrush fc = new(ForeColor))
            {
                if (Image == null)
                {
                    //Fix label maladjustment in center position
                    if (TextAlign == ContentAlignment.MiddleCenter)
                    {
                        if (Width % 2 != 0) TextAndImageRect.Offset(1, 0);
                        if (Height % 2 != 0) TextAndImageRect.Offset(0, 1);
                    }

                    G.DrawString(Text, Font, fc, TextAndImageRect, sf);
                }

                else if (string.IsNullOrWhiteSpace(Text))
                {
                    Rectangle imageRect = GetImageRectangle(TextAndImageRect, Image.Size, ImageAlign);

                    //Fix image maladjustment in center position
                    if (ImageAlign == ContentAlignment.MiddleCenter)
                    {
                        if (Width % Image.Width != 0) imageRect.Offset(1, 0);
                        if (Height % Image.Height != 0) imageRect.Offset(0, 1);
                    }

                    G.DrawImage(Enabled ? _image : _image.Grayscale(), imageRect);
                }

                else if (Image != null && !string.IsNullOrWhiteSpace(Text))
                {
                    if (TextImageRelation == TextImageRelation.Overlay)
                    {
                        Rectangle imageRect = GetImageRectangle(TextAndImageRect, Image.Size, ImageAlign);

                        //Fix image maladjustment in center position
                        if (ImageAlign == ContentAlignment.MiddleCenter)
                        {
                            if (Width % Image.Width != 0) imageRect.Offset(1, 0);
                            if (Height % Image.Height != 0) imageRect.Offset(0, 1);
                        }

                        //Fix label maladjustment in center position
                        if (TextAlign == ContentAlignment.MiddleCenter)
                        {
                            if (Width % 2 != 0) TextAndImageRect.Offset(1, 0);
                            if (Height % 2 != 0) TextAndImageRect.Offset(0, 1);
                        }

                        G.DrawImage(Enabled ? _image : _image.Grayscale(), imageRect);
                        G.DrawString(Text, Font, fc, TextAndImageRect, sf);
                    }

                    else
                    {
                        int innerSpacing = 1;
                        Size ImageSize = Image.Size;
                        SizeF TextSizeF = G.MeasureString(Text, Font, TextAndImageRect.Size, sf);
                        Size TextSize = new((int)TextSizeF.Width + innerSpacing * 2, (int)TextSizeF.Height);

                        Rectangle space = new();
                        Rectangle textRect = new();
                        Rectangle imageRect = new();

                        if (TextImageRelation == TextImageRelation.ImageAboveText)
                        {
                            space.Width = Math.Max(ImageSize.Width, TextSize.Width);
                            space.Height = ImageSize.Height + TextSize.Height + innerSpacing;

                            space = GetImageRectangle(TextAndImageRect, space.Size, ImageAlign);

                            imageRect = GetImageRectangle(space, Image.Size, ImageAlign);
                            imageRect.Y = space.Y;

                            textRect.Width = space.Width;
                            textRect.Height = TextSize.Height;
                            textRect.X = space.X;
                            textRect.Y = imageRect.Bottom + innerSpacing;
                        }

                        else if (TextImageRelation == TextImageRelation.TextAboveImage)
                        {
                            space.Width = Math.Max(ImageSize.Width, TextSize.Width);
                            space.Height = ImageSize.Height + TextSize.Height + innerSpacing;

                            space = GetImageRectangle(TextAndImageRect, space.Size, ImageAlign);

                            textRect.Width = space.Width;
                            textRect.Height = TextSize.Height;
                            textRect.X = space.X;
                            textRect.Y = space.Y;

                            imageRect = GetImageRectangle(space, Image.Size, ImageAlign);
                            imageRect.Y = textRect.Bottom + innerSpacing;
                        }

                        else if (TextImageRelation == TextImageRelation.ImageBeforeText)
                        {
                            space.Width = ImageSize.Width + TextSize.Width + innerSpacing;
                            space.Height = Math.Max(ImageSize.Height, TextSize.Height);

                            space = GetImageRectangle(TextAndImageRect, space.Size, ImageAlign);

                            imageRect = GetImageRectangle(space, ImageSize, ContentAlignment.MiddleLeft);

                            textRect.X = imageRect.Right + innerSpacing;
                            textRect.Y = space.Y + 1;
                            textRect.Width = TextSize.Width;
                            textRect.Height = space.Height;
                        }

                        else if (TextImageRelation == TextImageRelation.TextBeforeImage)
                        {
                            space.Width = ImageSize.Width + TextSize.Width + innerSpacing;
                            space.Height = Math.Max(ImageSize.Height, TextSize.Height);

                            space = GetImageRectangle(TextAndImageRect, space.Size, ImageAlign);

                            textRect.X = space.X;
                            textRect.Y = space.Y;
                            textRect.Width = TextSize.Width;
                            textRect.Height = space.Height;

                            imageRect.X = textRect.Right + innerSpacing;
                            imageRect.Y = space.Y;
                            imageRect.Width = ImageSize.Width;
                            imageRect.Height = ImageSize.Height;
                        }

                        G.DrawImage(Enabled ? _image : _image.Grayscale(), imageRect);
                        G.DrawString(Text, Font, fc, textRect, sf);
                    }
                }
            }

            #endregion

            base.OnPaint(e);


        }

        private Rectangle GetImageRectangle(Rectangle rect, Size size, ContentAlignment contentAlignment)
        {
            float CenterWidthF = rect.X + (float)(rect.Width - size.Width) / 2;
            float CenterHeightF = rect.Y + (float)(rect.Height - size.Height) / 2;

            int CenterWidth = (int)CenterWidthF;
            int CenterHeight = (int)CenterHeightF;

            return contentAlignment switch
            {
                ContentAlignment.TopLeft => new(rect.X, rect.Y, size.Width, size.Height),
                ContentAlignment.TopRight => new(rect.Right - size.Width, rect.Y, size.Width, size.Height),
                ContentAlignment.TopCenter => new(CenterWidth, rect.Y, size.Width, size.Height),
                ContentAlignment.MiddleLeft => new(rect.X, CenterHeight, size.Width, size.Height),
                ContentAlignment.MiddleCenter => new(CenterWidth, CenterHeight, size.Width, size.Height),
                ContentAlignment.MiddleRight => new(rect.Right - size.Width, CenterHeight, size.Width, size.Height),
                ContentAlignment.BottomLeft => new(rect.X, rect.Bottom - size.Height, size.Width, size.Height),
                ContentAlignment.BottomCenter => new(CenterWidth, rect.Bottom - size.Height, size.Width, size.Height),
                ContentAlignment.BottomRight => new(rect.Right - size.Width, rect.Bottom - size.Height, size.Width, size.Height),
                _ => new(CenterWidth, CenterHeight, size.Width, size.Height),
            };
        }

    }
}